### DISCUSSION TITLE
Analytical derivitives for SuperSimplex noise

### DISCUSSION THREAD

**Discussion by tayloraswift** [Razaekel/noise-rs#260] [Q&A] [answered]
Does anyone know how to compute the analytic derivative of SuperSimplex noise? The algorithm is so poorly documented I couldn’t find any information anywhere.

**Comment by mystise** [comment, marked as answer]
Checkout the eval_D2C1 and eval_D3C1 functions from SuperSimplexNoise.java, those provide what appears to be dx, dy, dxdy, and all the other various derivatives.

Edit: I just realized that gives you the numerical derivatives, not the analytical ones. I have the analytical equations, but the derivatives aren't super helpful unfortunately.

```
    // Calculation of maximum value:
    // x => real_rel_coords[0], y => real_rel_coords[1]
    // a-h, components of gradient vectors for 4 closest points
    // One contribution: (a*x + b*y) * (2/3 - x^2 - y^2)^4
    // Limit per contribution: 0 <= x^2 + y^2 < 2/3
    // skew = ((1 / sqrt(2 + 1) - 1) / 2)
    // (a*x + b*y) * (2/3 - x^2 - y^2)^4 + (c*(x - 1 - 2 * skew) + d*(y - 1 - 2 * skew)) * (2/3 - (x - 1 - 2 * skew)^2 - (y - 1 - 2 * skew)^2)^4 + (e*(x - skew) + f*(y - 1 - skew)) * (2/3 - (x - skew)^2 - (y - 1 - skew)^2)^4 + (g*(x - 1 - skew) + h*(y - skew)) * (2/3 - (x - 1 - skew)^2 - (y - skew)^2)^4
    // 0 <= x^2 + y^2 < 2/3 && 0 <= (x - 1 - 2 * skew)^2 + (y - 1 - 2 * skew)^2 < 2/3 && 0 <= (x - skew)^2 + (y - 1 - skew)^2 < 2/3 && 0 <= (x - 1 - skew)^2 + (y - skew)^2 < 2/3
    // a^2 + b^2 == 1 && c^2 + d^2 == 1 && e^2 + f^2 == 1 && g^2 + h^2 == 1
```

If you plug that into Mathematica you can take derivatives Dx and Dy, which end up as
![screen shot 2017-05-29 at 18 28 30](https://cloud.githubusercontent.com/assets/246232/26565061/adfe11c8-449c-11e7-883f-8359d3fee02c.png)
![screen shot 2017-05-29 at 18 28 37](https://cloud.githubusercontent.com/assets/246232/26565062/ae23c756-449c-11e7-94f3-fadb11e250f6.png)

**Comment by tayloraswift** [comment]
yikes

**Comment by mystise** [comment]
The derivatives for only one contribution (so you can just add them together after accounting for the different x and y positions, as in the eval_D2C1 and eval_D3C1 functions) are:

![screen shot 2017-05-29 at 23 05 48](https://cloud.githubusercontent.com/assets/246232/26570096/791d60e0-44c3-11e7-8cd9-d3e25dcddf95.png)

**Comment by tayloraswift** [comment]
are `x` and `y` relative to the gradient point

**Comment by mystise** [comment]
Yes
